### ISSUE TITLE
Dynamic linking with ASSERTIONS=2 and DISABLE_EXCEPTION_CATCHING=0 fails with "function in Table but not functionsInTableMap"

### ISSUE THREAD

**Issue by jprendes** [emscripten-core/emscripten#13347]
I am trying to build gtest as a side module, but I've managed to reduce it to the following code:

side.hpp
```c++
#pragma once
#include <vector>
std::vector<int> side(int n);
```

side.cpp
```c++
#include "side.hpp"
std::vector<int> side(int n) {
    std::vector<int> v;
    for (int i=0; i<n; i++) v.push_back(i);
    return v;
}
```

main.cpp
```c++
#include "side.hpp"
#include <stdio.h>
int main(void) {
    auto v = side(10);
    for (auto i : v)
        printf("%d\n", i);
    return 0;
}
```

I am building with the following commands:
```
em++ -s ASSERTIONS=2 -g -s DISABLE_EXCEPTION_CATCHING=0 -s SIDE_MODULE=1 -o side.wasm side.cpp
em++ -s ASSERTIONS=2 -g -s DISABLE_EXCEPTION_CATCHING=0 -s MAIN_MODULE=1 -s RUNTIME_LINKED_LIBS=['side.wasm'] -o main.html main.cpp
```

Running this aborts with the following stack trace
```
Assertion failed: function in Table but not functionsInTableMap
main.js:1435 Uncaught (in promise) RuntimeError: abort(Assertion failed: function in Table but not functionsInTableMap) at Error
    at jsStackTrace (http://127.0.0.1:8080/main.js:1707:19)
    at stackTrace (http://127.0.0.1:8080/main.js:2237:16)
    at abort (http://127.0.0.1:8080/main.js:1429:44)
    at assert (http://127.0.0.1:8080/main.js:598:5)
    at addFunctionWasm (http://127.0.0.1:8080/main.js:452:5)
    at updateGOT (http://127.0.0.1:8080/main.js:1783:34)
    at relocateExports (http://127.0.0.1:8080/main.js:1808:7)
    at postInstantiation (http://127.0.0.1:8080/main.js:1993:27)
    ...
```

**Comment by jprendes** [comment]
The problems arises from the extra check that ASSERTION=2 enables, verifying that `addFunctionWasm` isn't called with a function that has already been added to the table.

This is being hit because using `DISABLE_EXCEPTION_CATCHING=0` seems to force the compiler to add some methods to the side module's table AS WELL AS to the export list.

On instantiation the side module adds all the functions in its table to the main table and then `updateGOT` adds all the exported functions to the table.

In this particular case the first function being added for a second time is `_ZNSt3__26vectorIiNS_9allocatorIiEEE9push_backERKi`.

**Comment by jprendes** [comment]
I am able to reproduce this without using an STL container and without `DISABLE_EXCEPTION_CATCHING=0`.

side.hpp
```c++
#pragma once
class  MyClass {
  private:
    int data;
  public:
    __attribute__ ((__visibility__("hidden")))
    void set(int x) { data = x; }
    int & get() { return data; }
};
MyClass side(int n);
MyClass side(int n, void (MyClass::*fcn)(int));
```

side.cpp
```c++
#include "side.hpp"
MyClass side(int n) {
    return side(n, &MyClass::set);
}
// Force the compiler to add MyClass::set to the table
MyClass side(int n, void (MyClass::*fcn)(int)) {
    MyClass v;
    ((v).*(fcn))(n);
    return v;
}
```

main.cpp
```c++
#include "side.hpp"
#include <stdio.h>
int main(void) {
    auto v = side(42);
    printf("%d\n", v.get());
    return 0;
}
```

compiled with
```
em++ -s ASSERTIONS=2 -g -s SIDE_MODULE=1 -o side.wasm side.cpp
em++ -s ASSERTIONS=2 -g -s MAIN_MODULE=1 -s RUNTIME_LINKED_LIBS=['side.wasm'] -o main.html main.cpp
```

This aborts with due to `_ZN7MyClass3setEi`.

Tagging the method with `__attribute__ ((__visibility__("hidden")))` seems to be important here. Otherwise the function is not added to the table.
It seems that most (all?) the methods in the STL containers shipping with emscripten are tagged with the same attribute.

**Comment by jprendes** [comment]
It is definitely related with `__attribute__ ((__visibility__("hidden")))`.
Defining
```c++
#define _LIBCPP_HIDDEN __attribute__ ((__visibility__("default")))
```
before including the STL headers works around the problem on the STL example.

**Comment by donalffons** [comment]
I had the same error message in a simple "hello-world" application, using Embind, a main and a side module, `ASSERTIONS=2` and without `DISABLE_EXCEPTION_CATCHING=0`. I think those two issues are related. #13040

**Comment by jprendes** [comment]
> I think those two issues are related. #13040

I didn't see #13040 when I created this issue, my bad!

I gave it a go and tried adding `-D_LIBCPP_HIDDEN="__attribute__((__visibility__(\"default\")))"` the reproduction steps from #13040. That fixed the problem in there too, suggesting that they are definitely related if not the same issue.

I don't really know why the functions would be added to the table AND the exports, and I don't know if it's llvm, bynarien, or something else doing that. I also don't know why the STL in emscripten (musl?) is using that attribute, or what are the consequences of changing it.
